### ISSUE TITLE
Add module descriptor

### ISSUE THREAD

**Issue by XakepSDK** [ben-manes/caffeine#535]
Switch to module descriptor from automatic module name, so this library will have full support of java module system

**Comment by ben-manes** [comment]
Until Gradle 7.0, the build system wasn't compatible and external plugins like bnd were not either. That blocked prior attempts to add the descriptor.

The current blocker is that adding it in breaks Eclipse, which is where I do development. I don't know if this is a Gradle Buildship issue or Eclipse's. It is unable to handle a split project Imain, test, jmh).

I believe that the library's `module.java` would be,
```java
module com.github.benmanes.caffeine {
  exports com.github.benmanes.caffeine.cache.stats;
  exports com.github.benmanes.caffeine.cache;

  requires transitive org.checkerframework.checker.qual;
  requires transitive com.google.errorprone.annotations;
}
```

However until there is proper support in the toolchain I don't know how to include it. One approach would be package into the jar as a generated resource, which is what I've seen Maven users do, but that seems error prone.

**Comment by ben-manes** [comment]
If anyone wants to try and make this work a PR is welcome.

**Comment by XakepSDK** [comment]
Should annotations with runtime retention be transitive? Wouldn't be better to hide as much as possible?

**Comment by ben-manes** [comment]
`requires static` would be fine.

**Comment by ben-manes** [comment]
Released in 3.0.2 (thanks @A248!)

**Comment by soc** [comment]
I'm seeing build failures caused by this, as we explicitly exclude these dependencies in Maven:

> [ERROR] module not found: org.checkerframework.checker.qual
> [ERROR] module not found: com.google.errorprone.annotations


 @ben-manes suggested using `requires static`, but the patch by @A248 that was merged uses `requires`.

Could this be fixed?

**Comment by ben-manes** [comment]
Can you provide a reproduction? The static is in the descriptor. I’m not sure if this is a Maven bug or our misunderstanding.

https://github.com/ben-manes/caffeine/blob/b498408d9c36c2d29f75b1d73695f879639ae795/caffeine/src/main/java/module-info.java#L1-L7

**Comment by ben-manes** [comment]
This [explains](https://nipafx.dev/java-modules-optional-dependencies/) that it is needed during application compilation, but not runtime. That’s unlike Maven’s provided scope where the dependency needed it at compile time only but consumers did not. It seems there is no equivalent. Maybe you can change scoping rather than exclude?

**Comment by soc** [comment]
@ben-manes Oh, apologies, I scanned over the module-info incorrectly, the `static` is indeed there!

Well, I guess then that approach of avoiding the leak of those annotation packages into the project/IDE auto-import won't work naymore. :-(

**Comment by soc** [comment]
@ben-manes I'll give it a try, but I fear this doesn't provide the desired semantics ... we pretty much depend on annotation declarations being optional in the JVMS and not providing the jar file at all, but I think this is not something that is conceptually supported by Java modules.

**Comment by ben-manes** [comment]
I think you might need to have the exclusions and then also add the provided / optional dependencies explicitly with that scope setting. I don't think Maven's `exclusions` lets one change the scoping of a transitive, so you need to re-establish it at a lower scope. But maybe there is a neater way, e.g. some plugin that handles scope overriding.

**Comment by soc** [comment]
The problem is not at the level of Maven dependencies, but at the level of Java modules.
Getting fancy at the Maven level doesn't help if the Java module system simply doesn't accept it.

Both `requires` and `requires static` require the dependency to be there at compile-time.

**Comment by ben-manes** [comment]
Sorry, I kept thinking that this was a compilation failure due to excluding the dependency, rather than a runtime validation by the Java module system. If it is at runtime then there is no workaround.

**Comment by soc** [comment]
Yes, it's a compilation failure, but one based on the module system rules.

> If it is at runtime then there is no workaround.

Yes, I think this too.

Thanks for your thoughts and sorry for the noise, @ben-manes!

**Comment by ben-manes** [comment]
My naive expectation is that if you could scope the dependencies at `compileOnly` (which is what provided/optional do, iirc), then javac compilation should pass its module validation and at runtime `static` should pass by not requiring it on the module graph for class loading. Of course this is probably wrong or there are bugs since Java modules is so rarely used in practice.

**Comment by soc** [comment]
I think the problem here is that `compileOnly` doesn't provide the desired semantics: making it impossible that the dependency gets accidentally used by my code, that hasn't been modularized itself already.

**Comment by soc** [comment]
Btw, what's the reason for the `transitive`?

I just realized that not even the modularized parts of my code would be safe from these dependencies due to `transitive`.

**Comment by ben-manes** [comment]
You’re probably right it’s not needed. I probably had it above just trying to figure out the compatibility issues with Eclipse, before @A248 worked it out. I’m unsure though since it is annotations on the api and if removing would be great invalid if called through reflection?

What do you think @A248?

**Comment by soc** [comment]
I think it shouldn't have any influence on reflection, it only adds an invisible "requires com.google.errorprone.annotations..." to all modules that use caffeine.

**Comment by soc** [comment]
> using transitive is correct is because Caffeine exposes these annotation types

I don't get this ... with that logic, wouldn't all dependencies be declared `transitive`?

The only effect `transitive` has is that consuming modules don't need to write down the `requires` for it.

(I would argue that this is a misfeature, and should be avoided in pretty much any case.)

**Comment by A248** [comment]
> I don't get this ... with that logic, wouldn't all dependencies be declared transitive?
>
> The only effect transitive has is that consuming modules don't need to write down the requires for it.

`transitive` is not merely a feature of convenience, but the means of declaring a dependency which should be visible to API consumers, because a type from the dependency is exposed in the module's API.

This may help for explanation: https://nipafx.dev/java-modules-implied-readability/

Ignoring annotations for a moment, suppose you have the following method in an exported package. If `Bar` is from a dependency, `transitive` is a necessity here because API consumers will see the `Bar` return type:
```java
public Bar foo();
```

**Comment by soc** [comment]
> Ignoring annotations for a moment, suppose you have the following method in an exported package. If Bar is from a dependency, transitive is a necessity here because API consumers will see the Bar return type:

It really isn't – either the consumer refers to type `Bar`, then he can `require` it, or the consumer doesn't, then he can elide it.

I see nothing wrong with being explicit. If some module isn't required in the consumer's module-info file, I know it's not being used.

**Comment by A248** [comment]
Just FYI, use of `transitive` is further elaborated in the specification: https://openjdk.java.net/projects/jigsaw/spec/sotms/#implied-readability

Regardless, we'll receive helpful clarification from the jigsaw-dev mailing list with respect to handling annotation dependencies as used here. I've already sent them a message. Hopefully that will clear up any issue here.

**Comment by ben-manes** [comment]
@A248 I don't think your message went through. You have to subscribe to the mailing list to post.

**Comment by A248** [comment]
@soc You are right that we don't need to use `transitive` here.

The jigsaw-dev members (Alex Buckley and Remi Forax specifically) confirmed that `requires static`, without transitivity, is sufficient and works best. https://mail.openjdk.java.net/pipermail/jigsaw-dev/2021-June/thread.html

So, the caffeine module can be updated to remove the transitive declaration.

**Comment by ben-manes** [comment]
Thanks @A248. Let's make the change then. You're welcome to send a PR or I'll make the changes at some point soon.

**Comment by soc** [comment]
@ben-manes Does https://github.com/ben-manes/caffeine/pull/563 look as desired?

**Comment by ben-manes** [comment]
@soc lgtm

**Comment by A248** [comment]
All's good from what I see, for what it's worth.

**Comment by ben-manes** [comment]
fyi, there is a warning in the build that I am suppressing using `-Xlint:all,-exports`.

> warning: [exports] class Nullable in module org.checkerframework.checker.qual is not indirectly exported using requires transitive

**Comment by XakepSDK** [comment]
Should we ask igsaw-dev about this?

**Comment by ben-manes** [comment]
Sure, you are welcome to. They explicitly said annotations in the api don’t need to be transitive, so they might view this as ignorable or a bug. It’s certainly confusing and contradictory advice.

**Comment by A248** [comment]
I've mentioned this on jigaw-dev, so we'll find out what's happening here:
https://mail.openjdk.java.net/pipermail/jigsaw-dev/2021-June/014675.html

**Comment by hohonuuli** [comment]
This just bit me. A note for folks coming across the issue. The quickest fix is to downgrade from caffeine 3.0.2 to 3.0.1. 

Also, @ben-manes caffeine is awesome! Thanks for your continued work on it.

**Comment by ben-manes** [comment]
Released in 3.0.3

**Comment by nipafx** [comment]
[Alex Buckley replied to the thread.](https://mail.openjdk.java.net/pipermail/jigsaw-dev/2021-July/014682.html)

**Comment by soc** [comment]
Also, it seems that `requires static` without `transitive` does *not* break the Maven dependency exclusions, as I previously anticipated. So that's a pretty good outcome, I guess.

**Comment by A248** [comment]
> Released in 3.0.3

It might be valuable to briefly mention this in the 3.0.3 release notes.
